### PR TITLE
fix #97 Block not supported rule semantic

### LINKED ISSUE #97: Block not supported rule semantic
Currently we only allow:
Parent = Raw TS key that gets samples via TS.ADD
Child = Aggregated TS key using downsmapling rule

1. Each child can have 1 parent
2. A parent cannot have a rule that will aggregate to himself
3. A child cannot have any children
4. A parent can have multiple children

### PR DESCRIPTION
**Pull request by gkorland**

### CODE REVIEW

File: src/tsdb.c [hunk 1]
```diff
@@ -287,3 +284,41 @@ int SeriesHasRule(Series *series, RedisModuleString *destKey) {
...
     return FALSE;
 }
+
+int SeriesDeleteRule(Series *series, RedisModuleString *destKey) {
+    CompactionRule *rule = series->rules;
+	CompactionRule *prev_rule = NULL;
+	while (rule != NULL) {
+		if (RMUtil_StringEquals(rule->destKey, destKey)) {
+			if (prev_rule == NULL) {
+				free(series->rules);
```
Comment (gkorland) [inline src/tsdb.c]: Still a memory leak will be fixed in the delete fix

**Comment by gkorland** [comment]
ref #43

File: src/module.c [hunk 2]
```diff
@@ -677,29 +683,44 @@ int TSDB_createRule(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
+    }
+    RedisModule_RetainString(ctx, srcKeyName);
 
-    RedisModuleString *destKeyStr = RedisModule_CreateStringFromString(ctx, destKeyName);
-    if (SeriesAddRule(series, destKeyStr, aggType, bucketSize) == NULL) {
-    	return RedisModule_ReplyWithError(ctx, "ERROR creating rule");
+    // Last add the rule to source
+    destKeyName = RedisModule_CreateStringFromString(ctx, destKeyName);
+    if (SeriesAddRule(srcSeries, destKeyName, aggType, bucketSize) == NULL) {
+        RedisModule_ReplyWithSimpleString(ctx, "ERROR creating rule");
```
Comment (danni-m) [inline src/module.c]: TSDB: ....

File: src/module.c [hunk 3]
```diff
@@ -677,29 +683,44 @@ int TSDB_createRule(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
+        return REDISMODULE_ERR;
     }
-    RedisModule_RetainString(ctx, destKeyStr);
-    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_RetainString(ctx, destKeyName);
```
Comment (danni-m) [inline src/module.c]: why not copy the string inside SeriesAddRule and retain it there?

File: src/tsdb.c [hunk 4]
```diff
@@ -287,3 +298,41 @@ int SeriesHasRule(Series *series, RedisModuleString *destKey) {
     }
     return FALSE;
 }
+
+int SeriesDeleteRule(Series *series, RedisModuleString *destKey) {
+    CompactionRule *rule = series->rules;
+	CompactionRule *prev_rule = NULL;
```
Comment (danni-m) [inline src/tsdb.c]: somthing funky with spacing here.

File: src/tsdb.c [hunk 5]
```diff
@@ -287,3 +298,41 @@ int SeriesHasRule(Series *series, RedisModuleString *destKey) {
...
+	while (rule != NULL) {
+		if (RMUtil_StringEquals(rule->destKey, destKey)) {
+			if (prev_rule == NULL) {
```
Comment (danni-m) [inline src/tsdb.c]:
```suggestion
                         CompactionRule *next = rules = rule->nextRule;
                         FreeCompactionRule(series->rules);
			if (prev_rule != NULL) {
			     // cut off the current rule from the linked list
			     prev_rule->nextRule = next;
			}  else {
			    // make the next one to be the first rule
			    series->rules = next;
			}
```

File: src/module.c (hunk 3 above)
Comment (gkorland) [reply src/module.c]: cause I noticed we don't pass context to this functions

File: src/module.c (hunk 2 above)
Comment (gkorland) [reply src/module.c]: fixed

File: src/tsdb.c (hunk 4 above)
Comment (gkorland) [reply src/tsdb.c]: fixed

Comment (danni-m) [reply src/tsdb.c]: @gkorland doesn't seem so, look at line 306 & 307

**Comment by danni-m** [comment]
@gkorland LGTM, we need to fix the spacing.

**Comment by gkorland** [comment]
@danni-m which spacing?

**Pull request merged**